man/choice.5
.TH CHOICE 5 "12 September 1995"
.FB
.SH NAME
choice \- choice file for data entry
.SH SYNOPSIS
Layout of the \fIFirstBase\fP choice files
.SH DESCRIPTION
A \fIFirstBase\fP
choice file is a file that is used from \fIdbedit\fP(1) or
\fIdbvedit\fP(1) for providing very
strict control over data entry of a particular field. The idea is that the
user does not need to actually type out all of the input for a given
field, but rather chooses a selection whose corresponding meaning is
then inserted into the field.
.PP
This makes uniform data entry for fields with a set of known values
quite easy \(em and eases the typing required for long, duplicate entries.
.PP
A \fIFirstBase\fP choice file consists of a series of choices, one choice per line.
Each line is separated into three distinct areas by white space.
These three areas are: the choice, the meaning, and comments.
Anything past the second word (as defined by white space) is considered to
be a comment. This comment line is displayed during the choice input also.
.PP
For example, the following choice file could be used to make sure that
characters representing a color are installed into a field as opposed to
some garbage input:
.sp 1
.nf
.ft CW
1 Red The color red.
2 Blue The color blue.
3 Green The color green.
4 Yellow The color yellow.
5 Orange The color orange.
.ft
.fi
.PP
Note that if field has a default value, then a <RETURN> is acceptable at
the choice input screen. Otherwise, a selection must be made.
In this simplistic example, the only valid selections are 1 through 5.
.PP
There are also two other types of control lines that can appear
in a choice file that are used by \fIdbvedit\fP(1).
Choice file lines of the form \fB"XYZ":row,col\fP will create persistent
comments through all of the screens. Also, a line of the 
form \fB$HELP helpfile\fP will specify an \fIadditional\fP helpfile for use
during the actual choice input.
.PP
Fields that use
nested choice files can also be built. These choice fields will allow a
nested set of choice menus to be used for data entry. As the user goes
through the nested levels, the values are saved and the whole string
is returned to the user as the value of the field, with each individual
choice or attribute separated by a '|' marker.
.PP
A nested choice file is a file that contains a set of individual choice
menus linked together by the use of unique labeling. For example,
this mechanism allows a set of distinct Version Numbers to be picked from
for each of a list of Operating Systems.
Such a choice file would look like:
.sp 1
.nf
.ft CW
s Sun_Microsystems
d Digital_Equipment
a Apple

$s
1 3.4 Version 3.4
2 3.5 Version 3.5
3 4.0 Version 4.0
4 $INPUT

$d
1 2.1 Version 2.1
2 2.2 Version 2.2
3 2.3 Version 2.3

$a
1 9.1 Version 9.1
2 9.2 Version 9.2
3 9.3 Version 9.3
.ft
.fi
.PP
Note the use of the unique labels \fB$s\fP, \fB$d\fP and
\fB$a\fP. Each menu that is to be nested must be labeled. The label
consists of the actual typed choice, with a prepended \fB$\fP character.
.PP
Another feature of choice files is the ability to override any presented
choices, entering a new string of the users choice. The special meta-token
value, \fB$INPUT\fP \(em used in the example above \(em is used to force
entry of a non-standard choice value.
Additionally, by using the \fB$INPUT*\fP meta-token, these user-entered
choice values will be marked with an asterisk (*) at the end of the field.
.PP 1
Note that the sub screen display mechanism can be disabled by
using the type \fIC\fP for choice instead of \fIc\fP.
In this case, the field editing behaves like macro expansion,
i.e. all other choice field behavior will stay the same.
.sp 1
Under normal operation a selection of a value for a
choice field fails if the choice is undefined. However, \fIdbvedit\fP(1)
and \fIdbvemit\fP(1)
have the ability to add new choices defined by the user during record entry
to the proper choice file. To enable this feature, the AddChoice Flag (-A)
for the individual field must be set from \fIdbdbas\fP(1).
.sp 1
.SH SEE ALSO
dbedit(1), dbdbas(1), setup(5), xchoice(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/libedit/choice.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: choice.c,v 9.2 2001/02/16 19:15:28 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Choiceinput_sid[] = "@(#) $Id: choice.c,v 9.2 2001/02/16 19:15:28 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>
#include <vdict.h>

#include <sys/types.h>
#include <sys/stat.h>

static FILE *cfs;		/* the choice file itself */
extern fb_cnode *chead, *cptr;
char *choice_attribute = "|";
static char *CHOICE_INPUT = "$INPUT";
static char *CHOICE_INPUT_STAR = "$INPUT*";
static char *STAR = " *";
static char *MSG_CHOICE_INPUT = "Enter Full Value: ";
static char *MSG_ADD_CHOICE = "'%s' Not Found. Enter Full Value: ";
static char *MSG_HELP = "<CTL>-X=Cancel";
extern char *cdb_prompt_choicemsg;	/* prompt for choice message */
extern char *cdb_dbvedit_cho_ploc;	/* prompt location for choice level */
extern short int cdb_dbvedit_cho_pilength; /* length, dbvedit, choice level */
extern short int cdb_record_level;	/* record level toggle switch */
extern short int scanner;

static long seektop = 0;
static long seekbot = 0;

extern long cdb_seekstack[];
extern short int cdb_topstack;
static int choicepages;

static int r_choiceinput(fb_field *k, char *line, int new, int row, int col,
   char *inp, char *label, int e_size, int okend, char *helpfile);
static int choiceselect(fb_field *f, char *inp, char *line, int e_size);
static int fchoice(fb_field *f, char *label, int lastrow);
static int movefs(FILE *fs, char *label);
static int addchoice(char *label, char *value, fb_field *f);
static int getinchoice(char *p, char *msg, char *def, int size);

extern long fb_pop(void);
extern char *fb_underscore(char *p, int rep_blank);

extern char *cdb_T_PREV;
extern char *cdb_T_NPREV;

/* 
 * choiceinput - routine called from edit_field in dbedit.c 
 *   idea here is to display a list of choices.
 *   user picks one, choiceinput returns associated string.
 *
 *   multi-value fields implemented to simulate mutli menu selections.
 *   e_size is used only to limit input entered.
 */
 
   fb_choiceinput(k, line, new, row, col, e_size, okend)
      fb_field *k;
      char *line;
      int new, row, col, e_size, okend;
      
      {
         char inp[FB_MAXLINE], label[FB_MAXLINE], helpfile[FB_MAXNAME], *p;
         int st;
	 
         if (k->help == NULL || *k->help == NULL ||
	       (cfs = fopen(k->help, FB_F_READ)) == NULL){
	    fb_serror(FB_CANT_OPEN, k->help, NIL);
	    return(FB_ERROR);
	    }
         fb_cx_set_seekfile(k->help);
         fb_cx_set_seekpoint(0);
         seektop = seekbot = 0;
         fb_initstack();
	 fb_initcomment(cfs, helpfile);
	 fb_initcnode();
	 label[0] = NULL;
	 st = r_choiceinput(k,line,new,row,col,inp,label,e_size,
	    okend,helpfile);
	 if (st == FB_AOK){
	    sprintf(label, "$%s", inp);
	    fb_addcnode(line);
	    while (movefs(cfs, label)){
	       st = r_choiceinput(k,line,new,row,col,inp,label,e_size,
	          okend,helpfile);
	       if (st != FB_AOK)
	          break;
	       sprintf(label, "$%s", inp);
	       fb_addcnode(line);
	       }
	    }
	 if (st == FB_AOK){
	    cptr = chead;
	    strcpy(line, cptr->c_meaning);
	    for (cptr = chead->c_next; cptr; cptr = cptr->c_next) {
	       if (strlen(line) + strlen(cptr->c_meaning) + 1 < k->size){
		  strcat(line, choice_attribute);
		  /* replace any attribute characters with a dash */
		  for (p = cptr->c_meaning; *p; p++)
		     if (*p == *choice_attribute)
		        *p = CHAR_MINUS;
		  strcat(line, cptr->c_meaning);
		  }
	       }
	    }
	 fb_freecnode();
	 fb_freeccom();
	 fclose(cfs);
	 return(st);
      }

   static r_choiceinput(k,line,new,row,col,inp,label,e_size, okend, helpfile)
      fb_field *k;
      char *line, *label, *inp, *helpfile;
      int new, row, col, e_size, okend;
      
      {
	 int minc = 0, maxc, icol, ploc_row, ploc_col, t_minc, silentchoice;
         int st, c_st = 0;
	 char *cp, buf[FB_MAXLINE], cx_status[FB_MAXLINE];

         choicepages = -1;
         if (k->type == FB_CHOICE){
            silentchoice = 0;
	    ploc_row = cdb_t_lines;
	    ploc_col = 2;
	    if (cdb_dbvedit_cho_ploc != NULL){
	       strcpy(buf, cdb_dbvedit_cho_ploc);
	       cp = strchr(buf, ',');
	       *cp = NULL;
	       ploc_row = atoi(buf);
	       ploc_col = atoi(cp + 1);
	       }

            row = ploc_row;	/* allow HELP signal */
	    icol = (strlen(cdb_prompt_choicemsg) + ploc_col);
	    maxc = -cdb_dbvedit_cho_pilength;
	    c_st = fchoice(k, label, row - 2);
            strcpy(cx_status, CX_CHOICE_SELECT);
	    }
	 else{			/* silent choice (C) OR window choice */
            fb_move(cdb_t_lines, 1); fb_clrtoeol();
	    ploc_row = cdb_t_lines;
	    ploc_col = 2;
            fb_move(ploc_row, ploc_col); fb_clrtoeol();
            silentchoice = 1;
	    maxc = -(k->size);
	    icol = col;
            strcpy(cx_status, CX_SILENT_CHOICE_SELECT);
	    }
	 if (scanner == 1){
	    FB_PAUSE();
	    return(FB_ERROR);
	    }
	 inp[0] = NULL;
         /* always allow FB_DSIGNAL now */
	 icol = -icol;
	 if (k->idefault == NULL || strlen(k->idefault) == 0)
	    minc = 1;
	 else if (equal(k->idefault, cdb_T_PREV) && k->prev == NULL)
	    minc = 1;
	 else if (equal(k->idefault, cdb_T_NPREV) && k->prev == NULL)
	    minc = 1;
	 if (!cdb_record_level)
	    t_minc = 0;
	 else
	    t_minc = minc;
         /* test here whether D and dU belong */
         fb_cx_push_env("EXNP", cx_status, NIL); 	/* fld level */
         if (choicepages > 1)
            fb_cx_add_buttons("dU");
         if (t_minc == 0)
            fb_cx_add_buttons("D");
         if (helpfile[0] != NULL)
            fb_cx_add_buttons("H");
         if (label[0] == NULL)
            fb_cx_add_buttons("4");
	 for (st = FB_ERROR; st == FB_ERROR ;){
	    for (;;){
	       fb_infoline();
               fb_move(ploc_row, ploc_col);
               fb_printw(cdb_prompt_choicemsg);
               if (choicepages <= 1)
                  buf[0] = NULL;
               else
                  strcpy(buf, "<CTL>-F=Next Page, <CTL>-B=Prev Page, ");
               strcat(buf, "<CTL>-X=Abort");
               fb_scrhlp(buf);
               st = fb_input(-row, icol, maxc, t_minc,
                  FB_ALPHA, inp, FB_ECHO, -FB_OKEND, FB_CONFIRM);

               /*
                * this code is intended to make more keystrokes
                * work in more places, more transparent
                * - make ^D work like FB_END for non new records
                * - make FB_END be some FB_QSIGNAL interpreted outside
                */
               if (st == FB_END && okend != FB_OKEND)
                  st = FB_QSIGNAL;
               if (st == FB_DSIGNAL && !new)
                  st = FB_END;

	       if (!cdb_record_level && st == FB_DEFAULT)
		  if ((k->fld != NULL && strlen(k->fld)>0) || minc)
		     st = FB_ESIGNAL;

/*
*              removing this enables the FB_DSIGNAL to filter back up
*
*	       if (st == FB_DSIGNAL){
*		  if (!minc)
*		     st = FB_DEFAULT;
*		  else
*		     st = FB_ESIGNAL;
*		  }
*/
               if (st == FB_QHELP){
                  if (helpfile[0] != NULL)
                     fb_fhelp(helpfile);
                  else
                     fb_fhelp(k->help);
                  }
               else if (st == FB_PAGEUP || st == FB_BSIGNAL){
                  if (cdb_topstack != 0)
                     seektop = fb_pop();
                  }
               else if (st == FB_PAGEDOWN || st == FB_FSIGNAL){
                  if (c_st == EOF){
                     cdb_topstack = 0;
                     seektop = cdb_seekstack[cdb_topstack];
                     }
                  else{
                     fb_push(seektop);
                     seektop = seekbot;
                     }
                  }
               else
                  break;
               if (!silentchoice)
	          c_st = fchoice(k, label, row - 2);
               else
                  fb_local_display(1);
	       }
	    fb_trim(inp);
	    if (st == FB_AOK){
               if ((st = choiceselect(k, inp, line, e_size)) == FB_ERROR)
                 if (k->choiceflag != CHAR_A)
                    fb_serror(FB_MESSAGE, SYSMSG[S_ILLEGAL], NIL);
	       }
	    }
         fb_cx_pop_env();
	 return(st);
      }

/*
 *  choiceselect - selects a line from a choice file. 
 *     format of choice file is: choice-word meaning-word [comment-word ..]
 *     where words are runs of non white space.
 */

   static choiceselect(f, inp, line, e_size)
      fb_field *f;
      char *inp, *line;
      int e_size;
      
      {
         int j, withstar = 0, st;
	 char inbuf[FB_MAXLINE], word[FB_MAXLINE], buf[FB_MAXLINE];
	 
	 st = FB_ERROR;
         if (cdb_topstack != 0)
            fseek(cfs, cdb_seekstack[0], 0);
         else
            fseek(cfs, seektop, 0);
	 for (; fgets(inbuf, FB_MAXLINE, cfs) != NULL ;)
	    if (inbuf[0] != '#' && inbuf[0] != '"' && inbuf[0] != '$')
	       if ((j=fb_getword(inbuf, 1, word)) != 0 && equal(inp, word)){
		  fb_getword(inbuf, j, line);
		  if (equal(line, CHOICE_INPUT) || 
		      equal(line, CHOICE_INPUT_STAR)){
	             st = getinchoice(buf, MSG_CHOICE_INPUT, NIL, e_size);
		     if (st == FB_AOK){
			if (equal(line, CHOICE_INPUT_STAR))
			   withstar = 1;
		        strcpy(line, buf);
			if (withstar)
			   strcat(line, STAR);
			}
		     else
		        line[0] = NULL;
		     }
		  else
		     st = FB_AOK;
		  if (strlen(line) > f->size)
		     line[f->size] = NULL;
		  fb_underscore(line, 0);
		  break;
		  }
	 if (st == FB_ERROR && f->choiceflag == CHAR_A && chead == NULL){
	    sprintf(inbuf, MSG_ADD_CHOICE, inp);
	    st = getinchoice(buf, inbuf, " ", e_size);
	    if (st == FB_AOK){
	       fb_underscore(inp, 1);
	       fb_underscore(buf, 1);
	       st = addchoice(inp, buf, f);
	       if (st == FB_AOK){
	          fb_underscore(buf, 0);
	          strcpy(line, buf);
		  }
	       }
	    else
	       st = FB_ERROR;		/* in case it was FB_DEFAULT */
	    }
	 return(st);
      }

/*
 * fchoice - display a choice file
 */
 
   static fchoice(f, label, lastrow)
      fb_field *f;
      char *label;
      int lastrow;
      
      {
	 register int i, lc;
	 int eof, j, t, pt[5], k, width = 0, cwidth = 0;
	 char line[FB_MAXLINE], word[FB_MAXLINE];
	 
         fb_move(3, 1); fb_clrtobot();
         fseek(cfs, seektop, 0);
         fb_cx_set_seekpoint(seektop);
	 for (lc = 0; fgets(line, FB_MAXLINE, cfs) != NULL;){
	    if (line[0] == '$' || line[0] == FB_NEWLINE || line[0] == NULL)
	       break;
	    if (line[0] != '#' && line[0] != '"'){
	       j = fb_getword(line, 1, word);
	       k = strlen(word);
	       j = fb_getword(line, j, word);
	       k += strlen(word);
	       width = MAX(width, k);
	       j = fb_getword(line, j, word);
	       k = strlen(word);
	       cwidth = MAX(cwidth, k);
	       lc++;
	       }
	    }
	 pt[1] = 5; pt[2] = 15;
	 			
	 pt[3] = pt[2] + MAX(8, width) + 1; /* 8 = length of 'meaning ' */
	 pt[4] = pt[3] + MAX(8, cwidth) + 1; /* 8 = length of 'previous' */
	 sprintf(line, SYSMSG[S_S_BLANK], SYSMSG[S_CHOICE_TITLE]);
	 strcat(line, f->id);
	 if (strlen(label) > 0)
	    sprintf(line, "%s - %s", line, (label + 1));
	 fb_move(4,pt[1]);
	 fb_stand(line);
	 fb_move(6,pt[1]); fb_printw(SYSMSG[S_CHOICE]);
	 fb_move(6,pt[2]); fb_printw(SYSMSG[S_MEANING]);
	 fb_move(6,pt[3]); fb_printw(SYSMSG[S_COMMENT]);
	 /* movefs(cfs, label, seektop); */
         fseek(cfs, seektop, 0);
         j = 0;
         /* display a single page of choices */
         for (eof = 0, i = 8; i <= lastrow; ){
            if (fgets(line, FB_MAXLINE, cfs) == NULL){
               eof = 1;
               break;
               }
            if (line[0] == '$' || line[0] == FB_NEWLINE || line[0] == NULL){
               eof = 1;
               break;
               }
            if (line[0] == '"' || line[0] == '#' || line[0] == '$')
               continue;
            else{
               for (k=t=1;t <= 2 && (k=fb_getword(line, k, word)) != 0; ++t ){
                  fb_move(i, pt[t]);
                  if (t == 1)
                     fb_stand(word);
                  else
                     fb_printw(fb_underscore(word, 0));
                  }
               fb_move(i, pt[t]);
               line[k + cdb_t_cols - pt[3]] = NULL;
               fb_printw(FB_FSTRING, line+k);
               }
            i++;
            if (lc == ++j){
               eof = 1;
               break;
               }
            }
         fb_showcnode(pt[4]);		/* display cnodes -- suggest column */
         fb_showcomment();		/* display cnodes -- suggest column */
         seekbot = ftell(cfs);
         /* for return codes use EOF or FB_AOK */
         if (choicepages == -1){
            if (eof)
               choicepages = 1;
            else
               choicepages = 2;	/* rather, at least 2 */
            }
         if (eof)
            return(EOF);
         else
	    return(FB_AOK);
      }

   static movefs(fs, label)		/* maybe move file pointer to label */
      FILE *fs;
      char *label;
      
      {
	 char inbuf[FB_MAXLINE], word[FB_MAXLINE];

         fseek(fs, seektop, 0);
	 if (strlen(label) > 0)
	    for (; fgets(inbuf, FB_MAXLINE, fs) != NULL ;){
	       fb_getword(inbuf, 1, word);
	       if (equal(word, label)){
                  seektop = ftell(fs);
                  cdb_topstack = 0;
		  return(1);
                  }
	       }
	 return(0);
      }

   static addchoice(label, value, f)	/* physically add a choice */
      char *label, *value;
      fb_field *f;
      
      {
         int fd;
	 char buf[FB_MAXLINE];
	 struct stat sbuf;

         fb_fmessage("Adding Choice ...");
	 if ((fd = open(f->help, READWRITE)) < 0)
	    return(FB_ERROR);
	 fb_s_lock(fd, 1, f->help);
	 fb_sync_fd(fd);
	 fstat(fd, &sbuf);
	 lseek(fd, 0L, 2); 		/* seek to end of file */
	 sprintf(buf, "%s %s\n", label, value);
	 write(fd, buf, strlen(buf));
	 fb_s_unlock(fd, f->help);
	 close(fd);
	 fb_move(cdb_t_lines, 1); fb_clrtoeol();
	 return(FB_AOK);
      }

/*
 * getinchoice - duplicate the fb_getfilename() function with size option
 */

   static getinchoice(p, msg, def, size)
      char *p, *msg, *def;
      int size;
      
      {  
         register int i, st;
	 int msize;
      
	 fb_fmessage(msg);
	 if (strlen(def) == 0){
	    i = 1;
            fb_cx_push_env("X", CX_KEY_SELECT, NIL);
            }
	 else{
	    i = 0;
            fb_cx_push_env("DX", CX_KEY_SELECT, NIL);
            }
 	 msize = strlen(msg);
         fb_scrhlp(MSG_HELP);
	 st = fb_input(cdb_t_lines,msize+1, size, i, FB_ALPHA, p, FB_ECHO,
            FB_NOEND, FB_CONFIRM);
         fb_cx_pop_env();
	 fb_move(cdb_t_lines, 1); fb_clrtoeol(); fb_refresh();
	 if (st == FB_ABORT)
	    return(FB_ERROR);
	 else if (st == FB_DEFAULT)
	    strcpy(p,  def);
	 fb_trim(p);
	 return(st);
      }
